colexec: materialize whole vectors at once

yuzefovich opened the pull request:
**colexec: improve materializer unit tests**

This commit introduces a separate benchmark for the materializer as well
as improves the columnarize-materialize test to use random types. The
latter change allows us to remove another unit test because it became
redundant.

Release note: None

**colexec: materialize whole vectors at once**

Previously, the materializer would perform the conversion one row at
a time, one column at a time. This commit refactors the conversion
method so that it converts the desired number of elements at once and
introduces a helper that can convert the whole batches. The helper is
now used by the materializer and the default builtin operator.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50725)
<!-- Reviewable:end -->

yuzefovich commented:
```
name                       old time/op    new time/op    delta
ColumnarizeMaterialize-24    1.91ms ± 0%    1.74ms ± 1%   -8.90%  (p=0.000 n=7+10)

name                       old speed      new speed      delta
ColumnarizeMaterialize-24  83.9MB/s ± 0%  92.1MB/s ± 1%   +9.77%  (p=0.000 n=7+10)

name                       old alloc/op   new alloc/op   delta
ColumnarizeMaterialize-24     325kB ± 0%     359kB ± 0%  +10.22%  (p=0.000 n=9+10)

name                       old allocs/op  new allocs/op  delta
ColumnarizeMaterialize-24     21.3k ± 0%     21.3k ± 0%   +0.00%  (p=0.000 n=9+8)
```

yuzefovich committed:
colexec: improve materializer unit tests

This commit introduces a separate benchmark for the materializer as well
as improves the columnarize-materialize test to use random types. The
latter change allows us to remove another unit test because it became
redundant.

Release note: None

jordanlewis reviewed (commented):
Nice performance improvement!

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-:-MArOZ_NENsrnJ6i6AqV:b-a4w693)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/vec_to_datum_tmpl.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-MArOVpnDqdlT7e7r1RL:-MArOVpnDqdlT7e7r1RM:bnsiq6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e6a41e3901a2ae1df13024fc84b9d00c2469fb8/pkg/sql/colexec/vec_to_datum_tmpl.go#L142)):*
> ```Go
> // selected according to sel.
> // Note: len(converted) must be no less than length.
> func _VEC_TO_DATUM(
> ```

So much duplicated code in this file. Can you make it a type loop? It seems like you'd just have to template the part where you set the datum at converted[idx].


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/vec_to_datum_tmpl.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-MArOVpnDqdlT7e7r1RL:-MArkaqu4QLpSliCKEjj:b19eo76) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e6a41e3901a2ae1df13024fc84b9d00c2469fb8/pkg/sql/colexec/vec_to_datum_tmpl.go#L142)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

So much duplicated code in this file. Can you make it a type loop? It seems like you'd just have to template the part where you set the datum at converted[idx].
</blockquote></details>

I added another commit (that I will squash later) that reduces some of the duplication but abuses somewhat the way templated code snippets are handled right now - by assuming that a function named "_CONVERTER" of the correct signature is in the scope. I also checked that these "converter" functions are always fully inlined (the exceptions are decimals, dates, and UUIDs). I couldn't figure out a better way to propagate that converter function.

However, I'm not quite sure what you mean by "a type loop". Are you referring to introducing some "template info" struct that would contain this converter function and would be type specific, and then we would execute the template on the slice of these structs? If so, I think it will be of the same verbosity as the current things are, we would just move the verbosity from `_tmpl` into `_gen` file, and I would rather not do it at this point. Or did you have something else in mind?


<!-- Sent from Reviewable.io -->

jordanlewis commented:
> Are referring to introducing some "template info" struct that would contain this converter function and would be type specific, and then we would execute the template on the slice of these structs?

Yes, that's what I meant. I think it would be much less verbose because, in each of the cases, there is only a small amount of per-type code.

jordanlewis reviewed (commented):
Could you try rewriting this to use ordinary Go functions and see if they get inlined? Perhaps there's no advantage in this case to making them lexically inlined.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-:-MB-xPwME9PjzGJo7pCo:be4mixr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/vec_to_datum_tmpl.go, line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-MB-xeO69vMnh5b_iYDf:-MB-xeO74eCiY7G1jIKE:b84bi9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c06fb5136bab3070f4f87ab1781e8bb0c8ab6af/pkg/sql/colexec/vec_to_datum_tmpl.go#L195)):*
> ```Go
> 		bools := col.Bool()
> 		_CONVERTER := func(tupleIdx int) tree.Datum {
> 			if bools[tupleIdx] {
> ```

You can do `return tree.DBool(bools[tupleIdx])` here I believe.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: materialize whole vectors at once

Previously, the materializer would perform the conversion one row at
a time, one column at a time. This commit refactors the conversion
method so that it converts the desired number of elements at once and
introduces a helper that can convert the whole batches. The helper is
now used by the materializer and the default builtin operator.

Release note: None

yuzefovich reviewed (commented):
Updated, RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-:-MB17dSS7o21Vhd35nm8:b-rq6iuz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/vec_to_datum_tmpl.go, line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-MB-xeO69vMnh5b_iYDf:-MB14wNl1OllLM4_Hx9K:b6yykpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c06fb5136bab3070f4f87ab1781e8bb0c8ab6af/pkg/sql/colexec/vec_to_datum_tmpl.go#L195)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You can do `return tree.DBool(bools[tupleIdx])` here I believe.
</blockquote></details>

Not exactly but fixed.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
I got to thinking about this. One thing that seems potentially suspicious about this is the choice to store the batch as converted `tree.Datum` in columns, considering that we're going to then need to reach back into these columns and produce a materialized row. I think this is introducing unnecessary overhead, but not sure.

My alternative suggestion would be to store a slice of rows, and do your column at a time materialization into the rows, which will at that point be ready to return on the next call to `Next.

I also got to thinking about this problem and filed #50857 to avoid using Materializer altogether when possible.

jordanlewis commented on pkg/sql/colfetcher/cfetcher.go:
```diff
@@ -884,8 +884,10 @@ func (rf *cFetcher) pushState(state fetcherState) {
-func (rf *cFetcher) getDatumAt(colIdx int, rowIdx int, typ *types.T) tree.Datum {
-	return colexec.PhysicalTypeColElemToDatum(rf.machine.colvecs[colIdx], rowIdx, &rf.table.da, typ)
+func (rf *cFetcher) getDatumAt(colIdx int, rowIdx int) tree.Datum {
+	res := []tree.Datum{nil}
```
Does this cause an allocation? I don't think so but if it did that would be bad news.

jordanlewis reviewed (commented):
Besides the idea in the previous comment, LGTM

yuzefovich reviewed (commented):
It's an interesting idea, and I prototyped it, but it performs worse both in the microbenchmarks (there is about 10-15% hit) and some of the TPCH queries (here the hit is less noticeable, maybe like 3%, it could also be the noise). I think the main reason for the slowdown is that the amount of memory allocated increases just because `EncDatum`s take more space than `Datum`s. Another thing is that allocating whole W vectors, one vector at a time, with the original approach is noticeably faster that allocating N rows of W length, one row at a time (meaning that O(batch width) vs O(batch length) comparison). We could probably play around with allocating a big flat `[]EncDatum` slice in which datums from the same column are contiguous and then populate the output row in the materializer by choosing appropriate elements, but I'm not sure whether it would be more performant. Yet another thing is that `EncDatum`s (I think) must be reset between batches so that we can remove any encoding the caller added. I'll leave a TODO to possibly investigate this further.

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-:-MB7ivEZB0n27qLiJw_n:byy2oye)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colfetcher/cfetcher.go, line 888 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50725#-MAo9QelJYIVZkFpKlQN-r5-888:-MB7mA9w93X2bun2Kx4L:bgu0cdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d007776e5fdee4bd8c3219d9eb427814f4ad5646/pkg/sql/colfetcher/cfetcher.go#L888)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Does this cause an allocation? I don't think so but if it did that would be bad news.
</blockquote></details>

I think it doesn't, but I'm not sure how I would quickly verify that. I don't think this allocation is important because this method is used only with tracing enabled or to return a pretty error, and the comment on the method does say explicitly that it should not be used in hot paths, so I think it's ok to rely on our feelings here (even if they are wrong, it seems like no big deal).


<!-- Sent from Reviewable.io -->
